File: Javascript/parser.js
function Parser()
{
    var operators = [
        { sign: '+', precedence: 2 },
        { sign: '-', precedence: 2 },
        { sign: '*', precedence: 3 },
        { sign: '/', precedence: 3 },
        { sign: '(', precedence: 1 }
    ];
    
    var tokens = [];
    var opstack = [];
    var output = [];
    
    /**
     * Parses input using Reverse Polish Notation (Postfix).
     */
    this.parse = function(input)
    {
        tokens = input.split('').filter(elem => elem !== ' ');
        
        while(tokens.length > 0)
        {
            var token = extractToken(tokens);
    
            if(this.isOperand(token))
            {
                output.push(token);
            }
            else if (token === '(')
            {
                opstack.push(token);
            }
            else if (token === ')')
            {
                // Move all operators between () to output.
                while(opstack.length > 0)
                {
                    var currentOperator = opstack.pop();
                    if(currentOperator === '(')
                        break;
    
                    output.push(currentOperator);
                }
            }
            else if (this.isOperator(token))
            {
                var operator = this.getOperator(token);
                var currentOpstackOperator = this.getOperator(opstack[opstack.length - 1]);
                // Move all consecutive operators with higher precedence to output.
                while(opstack.length > 0 && currentOpstackOperator.precedence >= operator.precedence)
                {
                    opstack.pop();
                    output.push(currentOpstackOperator.sign);
                    currentOpstackOperator = this.getOperator(opstack[opstack.length - 1]);
                }
                opstack.push(operator.sign);
            }
            else
            {
                console.log("NOT SUPPORTED TOKEN: " + token);
            }
        }
    
        while(opstack.length > 0)
        {
            var value = opstack.pop();
            output.push(value);
        }
    
        return output.toString();
    }
    
    this.isOperand = function(token)
    {
        return token.match(/[0-9a-zA-Z]/i);
    }
    
    this.isOperator = function(token)
    {
        return operators.some(op => op.sign === token);
    }
    
    this.getOperator = function(token)
    {
        return operators.find(op => op.sign === token);
    }

    function extractToken(value)
    {
        var token = '';
        while(isNumber(value[0]))
        {  
            token += value[0];
            value.splice(0, 1);
        }

        // Not a number.
        if(token === '')
        {
            token += value[0];
            value.splice(0, 1);
        }

        return token;
    }

    function isNumber(value)
    {
        return value.match(/[0-9]/i);
    }
}

File: C++/FunctionGrapher/FunctionGrapher/src/Parser.cpp
#include "Parser.h"
#include <stdexcept>
#include <stdexcept>

Parser::Parser()
{

}

std::vector<std::string> Parser::Parse(std::string input)
{
	std::vector<char> tokens = Utils::SplitWithoutSpaces(input);
	std::vector<Operator> opstack;
	std::vector<std::string> output;

	while (tokens.size() > 0)
	{
		std::string token = Utils::ExtractToken(tokens);

		if (IsOperand(token))
		{
			output.push_back(token);
		}
		else if (token == "(")
		{
			opstack.push_back(GetOperator(token));
		}
		else if (token == ")")
		{
			// Move all operators between () to output.
			while (opstack.size() > 0)
			{
				Operator op = opstack[opstack.size() - 1];
				opstack.pop_back();
				if (op.GetSign() == '(')
					break;

				output.push_back(std::string(1, op.GetSign()));
			}
		}
		else if (IsOperator(token))
		{
			Operator op = GetOperator(token);
			if (opstack.size() > 0)
			{
				Operator currOperator = opstack[opstack.size() - 1];
				// Move all consecutive operators with higher precedence to output.
				while (currOperator.GetPrecedence() >= op.GetPrecedence())
				{
					opstack.pop_back();
					output.push_back(std::string(1, currOperator.GetSign()));
					if (opstack.size() == 0)
						break;

					currOperator = opstack[opstack.size() - 1];
				}
			}

			opstack.push_back(op);
		}
		else
		{
			throw std::invalid_argument("Unexpected token:" + token);
		}
	}

	while (opstack.size() > 0)
	{
		Operator value = opstack[opstack.size() - 1];
		opstack.pop_back();
		output.push_back(std::string(1, value.GetSign()));
	}

	return output;
}

bool Parser::IsOperator(std::string& token) const
{
	for (unsigned int index = 0; index < m_Operators.size(); index++)
	{
		if (token == std::string(1, m_Operators[index].GetSign()))
			return true;
	}

	return false;
}

Operator Parser::GetOperator(std::string& token)
{
	for (unsigned int index = 0; index < m_Operators.size(); index++)
	{
		if (token == std::string(1, m_Operators[index].GetSign()))
		{
			return m_Operators.at(index);
		}
	}

	throw std::invalid_argument("Token does not exist");
}


bool Parser::IsOperand(std::string& token) const
{
	std::regex operands("^[0-9a-zA-Z\.]*$");
	
	return std::regex_match(token, operands);
}